Sources/LambdaConverters.Wpf/TemplateSelector.cs

```diff
@@ -0,0 +1,109 @@
+﻿using System;
+using System.Diagnostics;
+using System.Windows;
+using System.Windows.Controls;
+using JetBrains.Annotations;
+
+namespace LambdaConverters
+{
+    /// <summary>
+    /// A factory class used to create lambda-based instances of the <see cref="DataTemplateSelector"/> class.
+    /// </summary>
+    public static class TemplateSelector
+    {
+        sealed class Selector<I> : System.Windows.Controls.DataTemplateSelector
+        {
+            readonly Func<TemplateSelectorArgs<I>, DataTemplate> selectFunction;
+
+            SelectorErrorStrategy ErrorStrategy { get; }
+
+            bool IsSelectFunctionAvailable { get; }
+
+            [Pure]
+            DataTemplate GetErrorValue(DataTemplate defaultValue)
+            {
+                switch (ErrorStrategy)
+                {
+                    case SelectorErrorStrategy.ReturnDefaultValue:
+                        return defaultValue;
+
+                    case SelectorErrorStrategy.ReturnNewEmptyDataTemplate:
+                        return new DataTemplate();
+
+                    default:
+                        throw new NotSupportedException();
+                }
+            }
+
+            internal Selector(
+                Func<TemplateSelectorArgs<I>, DataTemplate> selectFunction,
+                SelectorErrorStrategy errorStrategy)
+            {
+                ErrorStrategy = errorStrategy;
+                IsSelectFunctionAvailable = selectFunction != null;
+
+                this.selectFunction = selectFunction;
+            }
+
+            DataTemplate SelectTemplateInternal(object item, DependencyObject container)
+            {
+                I inputValue;
+                try
+                {
+                    inputValue = (I)item;
+                }
+                catch (SystemException e) when (e is InvalidCastException || e is NullReferenceException)
+                {
+                    EventSource.Log.UnableToCastToInputType(item?.GetType().Name ?? "null", typeof(I).Name, ErrorStrategy.ToString());
```

Comment:
Do you mean to call the `UnableToCastToItemType` method? If no, please remove the `UnableToCastToItemType` method.

Reply:
In the ValueConverter.cs UnableToCastToInputType is called after failed casts as well. I wanted to be consistent to it.

Reply:
Agree, but the event keyword would be `Converters` instead of `Selectors`. Should we use the `UnableToCastToItemType` method?

Reply:
I totally overlooked that. Thank you, for spotting it.